File: py/0207_CourseSchedule.py
# 207. Course Schedule

from utils import chunk
from typing import List
from collections import defaultdict, deque


class Solution:
    def canFinish(self, numCourses: int, prerequisites: List[List[int]]) -> bool:
        indegree = [0] * numCourses
        adjList = defaultdict(list)
        # b is the pre-requisite of a
        for a, b in prerequisites:
            indegree[a] += 1
            adjList[b].append(a)

        queue = deque()
        for i in range(numCourses):
            if indegree[i] == 0:
                queue.append(i)

        if len(queue) == 0:
            return False

        # bfs to number courses
        order = []
        while queue:
            curr = queue.popleft()
            order.append(curr)
            for neighbour in adjList[curr]:
                indegree[neighbour] -= 1
                if indegree[neighbour] == 0:
                    queue.append(neighbour)

        print(order)
        return sum(indegree) == 0


if __name__ == "__main__":
    testSize = 2
    puzzles = [
        2,
        [[1, 0]],
        2,
        [[1, 0], [0, 1]],
        5,
        [[1, 0], [2, 0], [3, 1], [3, 2], [4, 3]],
        3,
        [[1, 0], [1, 2], [0, 1]],
        5,
        [[1, 4], [2, 4], [3, 1], [3, 2]]
    ]
    for puzzle in chunk(puzzles, testSize):
        print(Solution().canFinish(*puzzle))

File: go/0207/0207_CourseSchedule.go
// 207. Course Schedule

package main

import (
	"bufio"
	"encoding/json"
	"fmt"
	"os"
	"strconv"

	"github.com/samber/lo"
)

func canFinish(numCourses int, prerequisites [][]int) bool {
	// we can perform dfs to see how many courses we can achieve from the starting courses
	var adjList = make([][]int, numCourses)
	for i := range numCourses {
		adjList[i] = []int{}
	}

	// inDegree is also the number of prerequisite courses of "want"
	var inDegree = make([]int, numCourses)
	for _, p := range prerequisites {
		want, need := p[0], p[1]
		adjList[need] = append(adjList[need], want)
		inDegree[want] += 1
	}

	queue := []int{}
	for start, in := range inDegree {
		if in == 0 {
			queue = append(queue, start)
		}
	}

	count := 0
	for len(queue) > 0 {
		curr := queue[0]
		queue = queue[1:]
		count += 1

		// kahn's algorithm: explore node only if all prerequisites are fulfilled
		for _, unlocked := range adjList[curr] {
			inDegree[unlocked] -= 1
			if inDegree[unlocked] == 0 {
				queue = append(queue, unlocked)
			}
		}
	}

	return count == numCourses
}

func main() {
	scanner := bufio.NewScanner(os.Stdin)
	var lines []string
	for scanner.Scan() {
		lines = append(lines, scanner.Text())
	}

	const testSize int = 2

	for _, test := range lo.Chunk(lines, testSize) {
		var prerequisites [][]int
		numCourses, _ := strconv.Atoi(test[0])
		json.Unmarshal([]byte(test[1]), &prerequisites)
		result := canFinish(numCourses, prerequisites)
		fmt.Println(result)
	}
}
